File: behavioral/command.py
from abc import ABC, abstractmethod
from typing import Optional


# Command Interface
class Command(ABC):
    @abstractmethod
    def execute(self) -> None:
        pass

    @abstractmethod
    def undo(self) -> None:
        pass


# Receiver: Light
class Light:
    def turn_on(self) -> None:
        print("Light is On")

    def turn_off(self) -> None:
        print("Light is Off")


# Concrete Command: LightOnCommand
class LightOnCommand(Command):
    def __init__(self, light: Light):
        self.light = light

    def execute(self) -> None:
        self.light.turn_on()

    def undo(self) -> None:
        self.light.turn_off()


# Concrete Command: LightOffCommand
class LightOffCommand(Command):
    def __init__(self, light: Light):
        self.light = light

    def execute(self) -> None:
        self.light.turn_off()

    def undo(self) -> None:
        self.light.turn_on()


# Invoker: RemoteControl
class RemoteControl:
    def __init__(self):
        self.command: Optional[Command] = None

    def set_command(self, command: Command) -> None:
        self.command = command

    def press_button(self) -> None:
        if self.command:
            self.command.execute()

    def press_undo(self) -> None:
        if self.command:
            self.command.undo()


# Example usage
light = Light()
light_on = LightOnCommand(light)
light_off = LightOffCommand(light)

remote = RemoteControl()

remote.set_command(light_on)
remote.press_button()  # Light is On
remote.press_undo()  # Light is Off

remote.set_command(light_off)
remote.press_button()  # Light is Off
remote.press_undo()  # Light is On

File: behavioral/command.cpp
#include <iostream>
#include <memory>

// Command Interface
class Command
{
public:
  virtual ~Command() = default;
  virtual void execute() = 0;
  virtual void undo() = 0;
};

// Receiver: Light
class Light
{
public:
  void turnOn()
  {
    std::cout << "Light is On" << std::endl;
  }

  void turnOff()
  {
    std::cout << "Light is Off" << std::endl;
  }
};

// Concrete Command: LightOnCommand
class LightOnCommand : public Command
{
public:
  LightOnCommand(Light *light) : light(light) {}

  void execute() override
  {
    light->turnOn();
  }

  void undo() override
  {
    light->turnOff();
  }

private:
  Light *light;
};

// Concrete Command: LightOffCommand
class LightOffCommand : public Command
{
public:
  LightOffCommand(Light *light) : light(light) {}

  void execute() override
  {
    light->turnOff();
  }

  void undo() override
  {
    light->turnOn();
  }

private:
  Light *light;
};

// Invoker: RemoteControl
class RemoteControl
{
public:
  void setCommand(std::shared_ptr<Command> command)
  {
    this->command = command;
  }

  void pressButton()
  {
    if (command)
    {
      command->execute();
    }
  }

  void pressUndo()
  {
    if (command)
    {
      command->undo();
    }
  }

private:
  std::shared_ptr<Command> command;
};

// Example usage
int main()
{
  auto light = std::make_shared<Light>();
  auto lightOn = std::make_shared<LightOnCommand>(light.get());
  auto lightOff = std::make_shared<LightOffCommand>(light.get());

  RemoteControl remote;

  remote.setCommand(lightOn);
  remote.pressButton(); // Light is On
  remote.pressUndo();   // Light is Off

  remote.setCommand(lightOff);
  remote.pressButton(); // Light is Off
  remote.pressUndo();   // Light is On

  return 0;
}
